aria/workflow_engine/engine.py

```diff
@@ -68,57 +68,83 @@ def create_workflow(
+                self._create_operation_task(operation_or_workflow, operation_dependencies)
+            else:
+                self.create_workflow(
+                    log_indentation_count=log_indentation_count + 1,
+                    graph=operation_or_workflow,
+                    workflow_context=workflow_context,
+                    start_task_cls=StartSubWorkflowTask,
+                    end_task_cls=EndSubWorkflowTask,
+                    depends_on=operation_dependencies)
+
+        workflow_tasks = self._workflow_tasks(graph.tasks)
+        # Every workflow has to have a dependency on the start of the workflow itself
+        workflow_tasks.append(workflow_start_task)
+
+        self.workflow_tasks[self._end_task_name(graph.name)] = type(
+            self._end_task_name(graph.name),
+            (end_task_cls,),
+            {})(context=workflow_context, depends_on=workflow_tasks)
+
+    def _create_operation_task(self, operation_context, depends_on):
         self.workflow_tasks[operation_context.name] = type(
             operation_context.name,
             (OperationTask,),
             {})(context=operation_context,
-                depends_on=self._operation_requires(graph, operation_context),
+                depends_on=depends_on,
                 **operation_context.engine_options)
 
-    def _task_is_operation(self, task):
+    def _get_tasks_from_dependencies(self, dependencies):
+        tasks = []
+        for task in dependencies:
+            task_name = (task.name
+                         if self._is_operation(task)
+                         else self._end_task_name(task.name))
+            tasks.append(self.workflow_tasks[task_name])
+        return tasks
+
+    def _get_dependencies_from_graph(self, graph, task):
+        dependencies = []
+        for dependent_task in graph.task_dependencies(task):
+            task_name = (dependent_task.name
+                         if self._is_operation(dependent_task)
+                         else self._end_task_name(dependent_task.name))
+            dependencies.append(self.workflow_tasks[task_name])
+        return dependencies
+
+    def _workflow_tasks(self, tasks):
+        workflow_tasks = []
+        for task in tasks:
+            task_name = (task.name
+                         if self._is_operation(task)
+                         else self._end_task_name(task.name))
+            workflow_tasks.append(self.workflow_tasks[task_name])
+        return workflow_tasks
+
+    def _is_operation(self, task):
         return isinstance(task, OperationContext)
 
-    def _operation_requires(self, graph, context):
-        return [
-            self.workflow_tasks[context_dependency.name]
-            for context_dependency in graph.task_dependencies(context)
-        ]
-
-    def _workflow_requires(self, contexts):
-        return [
-            self.workflow_tasks[context.name]
-            for context in contexts
-        ]
+    def _end_task_name(self, task_name):
```

Comment:
in the __init__ method we cal create this
 self._end_task_name = 'Ending.{0}'.format
self._start_task_name = 'Starting.{0}'.format